Refactor and add new tests to hugepages e2e tests

cynepco3hahue opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

Add e2e tests to cover usage of multiple hugepages with different
page sizes under the same pod.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:
It should be merged together with new prow jobs - https://github.com/kubernetes/test-infra/pull/17634

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [Usage]: https://github.com/kubernetes/enhancements/blob/8137e8370a55b789f17fbe5e674a690063a5b735/keps/sig-node/20190129-hugepages.md#pod-specification
```

Signed-off-by: Artyom Lukianov <alukiano@redhat.com>

cynepco3hahue commented:
I unsure what kind should I set.

Joseph-Irving commented:
/kind feature
/ok-to-test

cynepco3hahue commented:
/retest

cynepco3hahue commented:
/retest

cynepco3hahue commented:
/retest

bart0sh commented:
/lgtm

cynepco3hahue commented:
/test pull-kubernetes-node-kubelet-serial-hugepages

cynepco3hahue commented:
/test pull-kubernetes-node-kubelet-serial-hugepages

dims commented:
/assign @sjenning @derekwaynecarr

derekwaynecarr commented:
/approve
/lgtm

This is an improvement over the prior test by testing more than one page size.  I think we could restructure this test in the future to detect all huge page sizes supported, and attempt to use each page size discovered.

cynepco3hahue commented:
> This is an improvement over the prior test by testing more than one page size. I think we could restructure this test in the future to detect all huge page sizes supported, and attempt to use each page size discovered.

I totally agree it can be useful to test on top of different architectures.

bsdnet commented on test/e2e_node/hugepages_test.go:
```diff
@@ -90,145 +116,309 @@ func makePodToVerifyHugePages(baseName string, hugePagesLimit resource.Quantity)
 			}
-			ginkgo.By("configuring the host to reserve a number of pre-allocated hugepages")
+
+			ginkgo.By("Configuring the host to reserve a number of pre-allocated hugepages")
```
Same here if we can include size and count

cynepco3hahue replied:
done

bsdnet commented on test/e2e_node/hugepages_test.go:
```diff
@@ -33,30 +36,53 @@ import (
 
-	command := ""
+	hugetlbLimitFile := ""
 	// this command takes the expected value and compares it against the actual value for the pod cgroup hugetlb.2MB.<LIMIT>
```
hugetlb.2MB ==> e.g hugetlb.2MB to illustrate the format ?

cynepco3hahue replied:
Can you please clarify, I unsure that I understand the comment?

cynepco3hahue commented on test/e2e_node/hugepages_test.go:
```diff
@@ -90,145 +116,309 @@ func makePodToVerifyHugePages(baseName string, hugePagesLimit resource.Quantity)
-	err := exec.Command("/bin/sh", "-c", "echo 5 > /proc/sys/vm/nr_hugepages").Run()
-	if err != nil {
+
+	command := fmt.Sprintf("echo %d > %s-%dkB/%s", hugepagesCount, hugepagesDirPrefix, hugepagesSize, hugepagesCapacityFile)
```
done

cynepco3hahue commented on test/e2e_node/hugepages_test.go:
```diff
@@ -90,145 +116,309 @@ func makePodToVerifyHugePages(baseName string, hugePagesLimit resource.Quantity)
 	}
-	outData, err := exec.Command("/bin/sh", "-c", "cat /proc/meminfo | grep 'HugePages_Total' | awk '{print $2}'").Output()
+
+	command = fmt.Sprintf("cat %s-%dkB/%s", hugepagesDirPrefix, hugepagesSize, hugepagesCapacityFile)
```
done

cynepco3hahue commented on test/e2e_node/hugepages_test.go:
```diff
@@ -90,145 +116,309 @@ func makePodToVerifyHugePages(baseName string, hugePagesLimit resource.Quantity)
+	setHugepages := func() {
+		for hugepagesResource, count := range hugepages {
+			size := resourceToSize[hugepagesResource]
+			ginkgo.By("Verifying hugepages are supported")
```
done

cynepco3hahue commented on test/e2e_node/hugepages_test.go:
```diff
@@ -90,145 +116,309 @@ func makePodToVerifyHugePages(baseName string, hugePagesLimit resource.Quantity)
-	val, ok := node.Status.Capacity[v1.ResourceName(resourceName)]
-	if !ok {
-		return ""
+// isHugePageSupported returns true if hugepages of the specified size is available on the host
```
done

Commit:
Refactor and add new tests to hugepages e2e tests

Add tests to cover usage of multiple hugepages with different
page sizes under the same pod.

Signed-off-by: Artyom Lukianov <alukiano@redhat.com>

cynepco3hahue commented:
/test pull-kubernetes-node-kubelet-serial-hugepages

cynepco3hahue commented:
@bart0sh Can you please review again?

cynepco3hahue commented:
/retest

odinuge reviewed (approved):
Thanks for the cleanup! As other said, it would also be nice with support for other sizes, but this is a great start. Ref. https://github.com/kubernetes/kubernetes/pull/78495 arm64 support 4 different sizes. ~As far as I remember the CI infra on gce only support 2MiB huge pages, but that would only skip the 1GiB pages, so that should be no issue~.

I haven't run the tests locally, but the changes looks good to me. Since this is already approved, ill add a hold so it isn't merged right away. In case some other reviewer or you @cynepco3hahue would like to merge it, feel free to cancel the hold. :smile: 

/lgtm
/kind cleanup
/hold

cynepco3hahue commented:
@odinuge Thanks for the review, in case, if you want to test hugepages under the CI, we have now a separate job for it - `pull-kubernetes-node-kubelet-serial-hugepages, currently it supports 2Mi and 1Gi hugepages.

odinuge commented:
> You want to test hugepages under the CI, we have now a separate job for it - `pull-kubernetes-node-kubelet-serial-hugepages, currently it supports 2Mi and 1Gi hugepages.

Yes, saw that just after leaving the review! Fantastic!

Here is the testgridlink for those who are looking for it: https://testgrid.k8s.io/sig-node-kubelet#kubelet-serial-gce-e2e-hugepages :smile:

cynepco3hahue commented:
/test pull-kubernetes-node-kubelet-serial-hugepages

cynepco3hahue commented:
/retest

odinuge reviewed (approved):
Another lgtm from me, as this looks good! :smile: Agree that it is smart to move it to a NodeSpecialFeature so we don't "help" ci-kubernetes-node-kubelet-serial to flake due to memory pressure.

I think this is good to go, so no hold this time.

/lgtm

odinuge commented:
/priority backlog

odinuge commented:
/hold cancel

cynepco3hahue commented:
/retest
